File: 6.z-字形变换.py
#
# @lc app=leetcode.cn id=6 lang=python3
#
# [6] Z 字形变换
#
# runtime: 79.25% - memory - 94.26%
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1:
            return s

        gap = (numRows - 1) * 2

        ret = ""
        for i in range(numRows):
            start = 0
            if i == 0 or i == numRows - 1:
                idx = i
                while idx < len(s):
                    ret += s[idx]

                    start += 1
                    idx = i + start * gap
            else:
                idx = i
                ref = 0
                if idx < len(s):
                    ret += s[idx]
                else:
                    continue
                while True:
                    start += 1
                    ref = start * gap

                    idx_l = ref - i
                    if idx_l < len(s):
                        ret += s[idx_l]
                    else:
                        break

                    idx_r = ref + i

                    if idx_r < len(s):
                        ret += s[idx_r]
                    else:
                        break

        return ret
        

File: 6.z-字形变换.js
/*
 * @lc app=leetcode.cn id=6 lang=javascript
 *
 * [6] Z 字形变换
 */
/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
// runtime: 87.58% - memory - 40.91%
var convert = function(s, numRows) {
    if (numRows == 1) return s;

    var ret = "";
    var arrs = [];
    for (var i = 0; i < numRows; i++) {
        arrs.push([]);
    }

    var pointer = 0;
    for (var i = 0; i < s.length; i++) {
        arrs[pointer].push(s[i]);
        if (Math.floor(i / (numRows - 1)) % 2 == 0) {
            pointer += 1;
        } else {
            pointer -= 1;
        }
    }
    for (var i = 0; i < arrs.length; i++) {
        for (var j = 0; j < arrs[i].length; j++) {
            ret += arrs[i][j];
        }
    }

    return ret;

};

